### PR TITLE
Compact stack traces in the diag log so the bug-report buffer covers more events

### PR DESCRIPTION
**Pull request by mikelward**
## Summary

Realised reading a real bug report that the `300 of max 300` log lines claim is technically true and practically misleading: each routine `MqttPublisher: publish failed` call writes ~13 newline-separated lines (1 message + a 12-frame trace of Netty / executor noise — no app code), so a single 80-second storm of retries chews through a third of the buffer and shoves the relevant history out. The retained frames don't help — they're all framework internals.

This PR routes `DiagLog.log`'s Throwables through a new `compactStackTraceString` that walks the cause chain like `printStackTrace` does but keeps only the **top frame per Throwable**. Each Throwable in the chain costs a flat two lines (class + message, plus the top `at …`); the deeper frames aren't summarised with `... N more` either — that line is dead weight at 1 frame and the budget is the bottleneck. `last-crash.txt` still gets the unchanged full trace via `writeCrashLog` for uncaught exceptions, so post-mortem debugging loses nothing.

### Before / after, one `MqttPublisher: publish failed` event

Before — 14 lines:

```
05-17 06:53:09.396 W MqttPublisher: MQTT publish failed to mqtt://192.168.0.240:1883/...
com.hivemq.client.mqtt.exceptions.ConnectionFailedException: io.netty.channel.ConnectTimeoutException: ...
Caused by: io.netty.channel.ConnectTimeoutException: connection timed out: /192.168.0.240:1883
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe$1.run(Unknown Source:34)
	at io.netty.util.concurrent.PromiseTask.runTask(Unknown Source:15)
	at io.netty.util.concurrent.ScheduledFutureTask.run(Unknown Source:50)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(Unknown Source:0)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(Unknown Source:0)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SourceFile:9)
	at io.netty.channel.nio.NioEventLoop.run(Unknown Source:168)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(Unknown Source:41)
	at io.netty.util.internal.ThreadExecutorMap$2.run(Unknown Source:8)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(Unknown Source:2)
	at java.lang.Thread.run(Thread.java:1564)
```

After — 4 lines:

```
05-17 06:53:09.396 W MqttPublisher: MQTT publish failed to mqtt://192.168.0.240:1883/...
com.hivemq.client.mqtt.exceptions.ConnectionFailedException: io.netty.channel.ConnectTimeoutException: connection timed out: /192.168.0.240:1883
Caused by: io.netty.channel.ConnectTimeoutException: connection timed out: /192.168.0.240:1883
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe$1.run(Unknown Source:34)
```

~70 % reduction on this shape, ~3× more logical events fit in the same 300-line budget, the call-site frame is still preserved at the top of each cause. (The outer `ConnectionFailedException` arrives with zero frames after R8 — its `at …` line is correctly omitted.)

### Defensive guards mirroring `printStackTrace`

- **Cyclic cause chains** (`a.cause = b; b.cause = a`) terminate via an `IdentityHashMap` and emit a `[CIRCULAR REFERENCE: …]` marker. Without this a pathological Throwable would spin the calling thread before the executor submit even happens. (Codex P2 — addressed in 18dbcee7.)
- **Suppressed exceptions** surface as a one-line `Suppressed: <class>: <message>` per parent, so try-with-resources / `.use` close failures aren't silently dropped. Frames are dropped to keep the budget tight; the class + message is enough to spot. (Codex P2 — addressed in 18dbcee7.)

## Test plan

- [x] `./gradlew :app:testDebugUnitTest --tests "app.clothescast.diag.DiagLogTest"`
- [x] New `DiagLogTest` cases cover: starts-with class+message; one-frame cap with no `more` summary; multi-throwable cause chain with `Caused by` prefix; empty-stack edge case (R8-stripped HiveMQ wrappers); cycle protection; suppressed-as-one-line summary
- [ ] CI: JVM unit tests + Android debug build green
- [ ] On a build of this branch: share a bug report after a few MQTT failures, confirm each trace block is 3 lines (class+message, `Caused by` line, `at …`) and that more pre-failure history is retained

https://claude.ai/code/session_01QiXz6VZ3Qg9SZJBdtRLaea

### CODE REVIEW

File: app/src/main/kotlin/app/clothescast/diag/DiagLog.kt [hunk 1]
```diff
@@ -182,12 +192,42 @@ object DiagLog {
...
+            current.message?.let { sb.append(": ").append(it) }
+            val frames = current.stackTrace
+            val keep = minOf(maxFrames, frames.size)
+            for (i in 0 until keep) {
+                sb.append('\n').append("\tat ").append(frames[i])
+            }
+            if (frames.size > keep) {
+                sb.append('\n').append("\t... ").append(frames.size - keep).append(" more")
+            }
+            current = current.cause
```
[the remark this answers is not in the document]
Comment (mikelward) [reply app/src/main/kotlin/app/clothescast/diag/DiagLog.kt:237]: Addressed in 18dbcee7. `compactStackTraceString` now tracks already-seen Throwables in an `IdentityHashMap` and bails with a `[CIRCULAR REFERENCE: …]` marker when the chain loops back — mirrors what `printStackTrace` does. Test added: `compactStackTraceString terminates on a circular cause chain` exercises the `a.cause = b; b.cause = a` shape.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (mikelward) [reply app/src/main/kotlin/app/clothescast/diag/DiagLog.kt:237]: Addressed in 18dbcee7. `compactStackTraceString` now iterates `current.suppressed` for each Throwable in the chain and emits a one-line `Suppressed: <class>: <message>` per entry — no frames (keeps the snapshot budget tight) but the class + message is enough to spot a try-with-resources / `.use` close failure that would otherwise vanish. Test added: `compactStackTraceString surfaces suppressed exceptions as a one-line summary`.

---
_Generated by [Claude Code](https://claude.ai/code)_
